### DISCUSSION TITLE
What the the practice for cleaning up cancelled CancellationTokenSource objects?

### DISCUSSION THREAD

**Discussion by aaronatweta** [Cysharp/UniTask#602] [Q&A]
Can someone tell me what the correct practice is for cleaning up a UniTask that has been cancelled via a CancellationTokenSource? In the UniTaskViewer panel I am seeing some dangling tasks with the "Cancelled" status so I want to know what needs to be done to make sure these things are cleaned up correctly?

I have seen exampled that have a Cancel, Dispose, New' idiom but that can't always be done if something else is cancelling the CTS such as the GetCancellationTokenOnDestroy or something that is cancelled by someone else.

**Comment by SERAP-KEREM** [comment]
My understanding is that a cancelled UniTask is already in a terminal state, so there isn't usually an additional cleanup step required for the UniTask itself.

The part that typically needs cleanup is the `CancellationTokenSource`, but only if you created and own it. In that case, the usual pattern is:

```csharp
cts.Cancel();
cts.Dispose();
```

(and create a new instance if you need to reuse the workflow).

If the token comes from something like `GetCancellationTokenOnDestroy()` or `destroyCancellationToken`, I wouldn't expect you to dispose anything yourself since the lifetime is managed elsewhere.

One thing I'm not completely sure about is whether the entries you're seeing in UniTaskTracker are actually still alive, or if the tracker is simply showing completed/cancelled tasks for diagnostic purposes. If they're in the `Cancelled` state rather than `Pending`, that sounds more like task history than an active leak.

I'd be more concerned if tasks remain `Pending` after the owning object has been destroyed or after cancellation has propagated.

Could you share whether those entries stay as `Cancelled` indefinitely, or whether they are actually `Pending`? A screenshot of the tracker might also help narrow it down.
